### PR TITLE
chore: ignore generated files

### PR DESCRIPTION
**Pull request by jeremybarbet**
# Remove all generated files from the repository

- [x] fix CI broken
- [x] move open API generated files to the script as well
- [x] update documentation
- [x] fix ts-node

## What

- Removing all generated files (schema, types definitions, fragment types, open API, **except contentfulTypes.d.ts**) from git and generating them locally on post-install/after pulling changes.

- `contentfulTypes.d.ts` will still be part of the GitHub repository. This [PR](https://github.com/island-is/island.is/pull/1144) removed the need to have a local `CONTENTFUL_MANAGEMENT_ACCESS_TOKEN` to generate the contentful types definitions and is now done on a private build step. Because of the external dependency on contentful, we want to keep the file on Github to avoid breaking our whole build pipeline in case contentful is down for some reason.

I had to do some changes here and there to make all schemas/types automatically _generatable_

**All projects using `autoSchemaFile` on the GraphQLModule:**
- Exported all resolvers to be able to generate the schema file with the script

**All projects using open API:**
- Exported the `DocumentBuilder()` into a separate file to import it for the buildOpenApi script

**air-discount-scheme-backend/application-system-api**
- Added an `INIT_SCHEMA` to ignore Redis initialization when used. Otherwise, the process keeps ending and never stop when building the open API schema

**content-search**
- I moved the dto/enums folder form the sub-level folder to the root folder (libs/api/domains/content-search/src/enums -> libs/api/content-search/src/enums). It was creating circular references when running the build schema script

- You can use `EXCLUDES_SCHEMAS=air-discount-scheme-api,gjafakort-api,skilavottord-ws` if you are not interested in building schema/types for some specific projects
- I also removed the HelloWorld module

## Why

All theses generated files creates a lot of noise in PR reviews and add more code owners on each PR for no real reason

## Screenshots / Gifs

— 

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] Formatting passes locally with my changes
- [x] I have rebased against master before asking for a review

### CODE REVIEW

File: libs/infra-nest-server/src/lib/execShellCommand.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import { logger } from '@island.is/logging'
+import { exec } from 'child_process'
+
+export const execShellCommand = (cmd: string): Promise<string> =>
+  new Promise((resolve, reject) => {
+    exec(cmd, (error, stdout, stderr) => {
+      if (error) {
+        console.log(error)
```
Comment (eirikurn) [inline libs/infra-nest-server/src/lib/execShellCommand.ts:8]: Should use logger

File: libs/logging/src/lib/logging.ts [hunk 2]
```diff
@@ -18,5 +18,5 @@ if (process.env.NODE_ENV === 'production') {
 export const logger = createLogger({
   level: logLevel,
   format: logFormat,
-  transports: [new transports.Console(), new SentryTransport()],
+  // transports: [new transports.Console(), new SentryTransport()],
```
Comment (eirikurn) [inline libs/logging/src/lib/logging.ts:21]: What is this for?

File: libs/infra-nest-server/src/lib/generateSchema.ts [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    })
+  }
+}
+
+const run = async () => {
+  logger.info('Running generate schema script...')
+  logger.info('Generating graphql schema files...')
+
+  const excludes = GENERATE_EXCLUDES_SCHEMA?.replace(/\s/g, '')?.split(',')
+  console.log('-excludes', excludes)
```
Comment (eirikurn) [inline libs/infra-nest-server/src/lib/generateSchema.ts:69]: logger? Need to add eslint error for console.log.

File: scripts/ts-node.js [hunk 4]
```diff
@@ -9,7 +9,13 @@
...
 //      yarn ts-node
 
+require('dotenv').config()
 require('ts-node').register({
-  compilerOptions: { module: 'commonjs', esModuleInterop: true },
+  compilerOptions: {
+    module: 'commonjs',
+    esModuleInterop: true,
+    noUnusedLocals: false,
+    strict: false,
```
Comment (eirikurn) [inline scripts/ts-node.js:18]: Why are these required? I'd think it should pick up the correct tsconfig of the executing file.

File: workspace.json [hunk 5]
```diff
@@ -345,6 +345,16 @@
             "buildTarget": "api:build"
           }
         },
+        "generate-schema": {
+          "builder": "@nrwl/workspace:run-commands",
+          "options": {
+            "commands": [
+              {
+                "command": "yarn ts-node apps/api/generateSchema.ts"
```
Comment (eirikurn) [inline workspace.json:353]: Does this file exist?

File: workspace.json [hunk 6]
```diff
@@ -3142,6 +3152,34 @@
           }
         }
       }
+    },
+    "generate-schema": {
+      "root": "libs/generate-schema",
```
Comment (eirikurn) [inline workspace.json:3157]: Does this lib exist?

**Review by eirikurn — CHANGES_REQUESTED** [review]
infra-nest-server should only be a consumable library for nest servers and not depend on all other libs (will cause nx circular dependency warnings).

I'd prefer each app/lib to define their own generate-schema and codegen scripts and use nx to run all of them `postinstall`.

Check out how `infra-nest-server/bootstrap` generates an openapi schema based on a CLI argument. We can do the same for graphql apps if they pass in an array of resolvers.

**Comment by eirikurn** [comment]
(sorry I looked. Know this is wip, but wanted to sneak in some implementation suggestions. Reach out if you wanna chat)

**Comment by jeremybarbet** [comment]
> infra-nest-server should only be a consumable library for nest servers and not depend on all other libs (will cause nx circular dependency warnings).
> 
> I'd prefer each app/lib to define their own generate-schema and codegen scripts and use nx to run all of them `postinstall`.
> 
> Check out how `infra-nest-server/bootstrap` generates an openapi schema based on a CLI argument. We can do the same for graphql apps if they pass in an array of resolvers.

Sounds very good, I wasn't sure where to include it yet, it was more to get something working first. Thanks for pointing out the direction. 

The rest is very much wip and messy so most of it will be reverted/deleted once everything is working properly, I will keep you updated

File: package.json [hunk 7]
```diff
@@ -20,6 +20,7 @@
     "affected:lint": "nx affected:lint",
     "affected:dep-graph": "nx affected:dep-graph",
     "affected:codegen": "nx affected --target=codegen",
+    "affected:schemas": "nx affected --target=init-schema --parallel --maxParallel=5 --exclude=$(grep EXCLUDES_SCHEMAS .env | cut -d '=' -f2)",
```
Comment (jeremybarbet) [inline package.json:23]: Let's say I'm only working on api/web project, I can do `EXCLUDES_SCHEMAS=air-discount-scheme-api,gjafakort-api,skilavottord-ws`. It saves a few seconds

File: package.json [hunk 8]
```diff
@@ -34,7 +35,9 @@
     "dep-graph": "nx dep-graph",
     "help": "nx help",
     "storybook:island-ui": "nx run island-ui-storybook:storybook",
-    "ts-node": "node -r ./scripts/ts-node"
+    "ts-node": "node -r ./scripts/ts-node",
+    "schemas": "test -f libs/api/schema/src/lib/schema.d.ts || echo 'export default () => {}' >> libs/api/schema/src/lib/schema.d.ts && yarn nx run api:contentful-types && yarn affected:schemas",
```
Comment (jeremybarbet) [inline package.json:39]: That's a bit ugly but the thing is `libs/api/schema/src/lib/schema.d.ts` is ignored, even though, without it, we can't build the API, because the file is required for some [import](https://github.com/island-is/island.is/blob/master/libs/api/schema/src/index.ts#L1). So if the file doesn't exist I created it with an empty function, and it gets overwritten once the codegen is done.

I will tackle that issue in another PR to remove this file and move all schema locally to each apps, so we won't depend on shared schemas for the apps. It already has been done for the [web project](https://github.com/island-is/island.is/blob/master/apps/web/codegen.yml)

File: .github/workflows/build.yml [hunk 9]
```diff
@@ -152,6 +152,22 @@ jobs:
...
         run: echo "::set-env name=BRANCH::${GIT_BRANCH}"
 
+      - name: Cache for generated files
+        id: generated-files-cache
+        uses: actions/cache@v2
+        with:
+          path: generated_files.tar.gz
+          key: ${{ runner.os }}-${{ hashFiles('libs/api/domains/cms/src/lib/generated/contentfulTypes.d.ts', 'apps/**/models/**/*.tsx?', 'libs/**/models/**/*.tsx?', 'apps/**/dto/**/*.tsx?', 'libs/**/dto/**/*.tsx?', 'apps/**/enums/**/*.tsx?', 'libs/**/enums/**/*.tsx?', 'apps/**/queries/**/*.tsx?', 'libs/**/queries/**/*.tsx?', 'apps/air-discount-scheme/web/screens/**/*.tsx', 'apps/air-discount-scheme/web/i18n/withLocale.tsx', 'apps/air-discount-scheme/web/components/AppLayout/AppLayout.tsx', 'apps/air-discount-scheme/web/components/Header/Header.tsx', 'apps/gjafakort/api/src/**/typeDefs.ts') }}-generated-files
+
+      - name: Debug hashkey
```
Comment (sindrig) [inline .github/workflows/build.yml:162]: Remove debug

File: .github/workflows/build.yml [hunk 10]
```diff
@@ -232,13 +248,28 @@ jobs:
...
           path: node_modules
           key: ${{ runner.os }}-${{ hashFiles('yarn.lock') }}-yarn
 
+      - name: Cache for generated files
+        uses: actions/cache@v2
+        with:
+          path: generated_files.tar.gz
+          key: ${{ runner.os }}-${{ hashFiles('libs/api/domains/cms/src/lib/generated/contentfulTypes.d.ts', 'apps/**/models/**/*.tsx?', 'libs/**/models/**/*.tsx?', 'apps/**/dto/**/*.tsx?', 'libs/**/dto/**/*.tsx?', 'apps/**/enums/**/*.tsx?', 'libs/**/enums/**/*.tsx?', 'apps/**/queries/**/*.tsx?', 'libs/**/queries/**/*.tsx?', 'apps/air-discount-scheme/web/screens/**/*.tsx', 'apps/air-discount-scheme/web/i18n/withLocale.tsx', 'apps/air-discount-scheme/web/components/AppLayout/AppLayout.tsx', 'apps/air-discount-scheme/web/components/Header/Header.tsx', 'apps/gjafakort/api/src/**/typeDefs.ts') }}-generated-files
+
+      - name: Debug hashkey
```
Comment (sindrig) [inline .github/workflows/build.yml:265]: ditto

File: .github/workflows/build.yml [hunk 11]
```diff
@@ -292,6 +324,20 @@ jobs:
...
```
Comment (sindrig) [inline .github/workflows/build.yml:333]: ditto

File: .github/workflows/build.yml [hunk 12]
```diff
@@ -311,13 +357,29 @@ jobs:
...
       - uses: actions/checkout@v2
         if: ${{ matrix.projects != 'dummy' }}
+
+      - name: Cache for generated files
+        uses: actions/cache@v2
+        with:
+          path: generated_files.tar.gz
+          key: ${{ runner.os }}-${{ hashFiles('libs/api/domains/cms/src/lib/generated/contentfulTypes.d.ts', 'apps/**/models/**/*.tsx?', 'libs/**/models/**/*.tsx?', 'apps/**/dto/**/*.tsx?', 'libs/**/dto/**/*.tsx?', 'apps/**/enums/**/*.tsx?', 'libs/**/enums/**/*.tsx?', 'apps/**/queries/**/*.tsx?', 'libs/**/queries/**/*.tsx?', 'apps/air-discount-scheme/web/screens/**/*.tsx', 'apps/air-discount-scheme/web/i18n/withLocale.tsx', 'apps/air-discount-scheme/web/components/AppLayout/AppLayout.tsx', 'apps/air-discount-scheme/web/components/Header/Header.tsx', 'apps/gjafakort/api/src/**/typeDefs.ts') }}-generated-files
+
+      - name: Debug hashkey
```
Comment (sindrig) [inline .github/workflows/build.yml:367]: ditto

File: .github/workflows/build.yml [hunk 13]
```diff
@@ -358,13 +420,29 @@ jobs:
...
```
Comment (sindrig) [inline .github/workflows/build.yml:430]: ditto

File: .github/workflows/build.yml [hunk 14]
```diff
@@ -405,13 +483,29 @@ jobs:
...
```
Comment (sindrig) [inline .github/workflows/build.yml:493]: ditto

File: .gitignore [hunk 15]
```diff
@@ -50,3 +50,22 @@ helm/**/*.tgz
...
+apps/gjafakort/web/graphql/possibleTypes.json
+apps/air-discount-scheme/api.graphql
+apps/air-discount-scheme/web/graphql/schema.tsx
+apps/air-discount-scheme/web/graphql/possibleTypes.json
+apps/air-discount-scheme/backend/src/openapi.yaml
+apps/skilavottord/ws/src/app/api.graphql
+apps/judicial-system/api/src/openapi.yaml
+apps/application-system/api/src/openapi.yaml
+apps/reference-backend/src/openapi.yaml
+# libs/api/domains/cms/src/lib/generated/contentfulTypes.d.ts
```
Comment (sindrig) [inline .gitignore:70]: confusing to keep this here, remove?

File: package.json [hunk 16]
```diff
@@ -20,6 +20,7 @@
     "affected:lint": "nx affected:lint",
     "affected:dep-graph": "nx affected:dep-graph",
     "affected:codegen": "nx affected --target=codegen",
+    "affected:schemas": "nx affected --target=init-schema --parallel --maxParallel=5 $(test -f .env && echo \"--exclude=$(grep EXCLUDES_SCHEMAS .env | cut -d '=' -f2)\")",
```
Comment (sindrig) [inline package.json:23]: When and how is the `EXCLUDE_SCHEMAS` used?

File: .github/workflows/build.yml [hunk 17]
```diff
@@ -405,13 +483,29 @@ jobs:
     steps:
       - uses: actions/checkout@v2
         if: ${{ matrix.projects != 'dummy' }}
+
+      - name: Cache for generated files
+        uses: actions/cache@v2
+        with:
+          path: generated_files.tar.gz
+          key: ${{ runner.os }}-${{ hashFiles('libs/api/domains/cms/src/lib/generated/contentfulTypes.d.ts', 'apps/**/models/**/*.tsx?', 'libs/**/models/**/*.tsx?', 'apps/**/dto/**/*.tsx?', 'libs/**/dto/**/*.tsx?', 'apps/**/enums/**/*.tsx?', 'libs/**/enums/**/*.tsx?', 'apps/**/queries/**/*.tsx?', 'libs/**/queries/**/*.tsx?', 'apps/air-discount-scheme/web/screens/**/*.tsx', 'apps/air-discount-scheme/web/i18n/withLocale.tsx', 'apps/air-discount-scheme/web/components/AppLayout/AppLayout.tsx', 'apps/air-discount-scheme/web/components/Header/Header.tsx', 'apps/gjafakort/api/src/**/typeDefs.ts') }}-generated-files
```
Comment (darrikonn) [inline .github/workflows/build.yml:491]: Would like to see a variable for this, which can be reused for all the keys. Something like:
```yaml
cache_key: &cache_key "hashFiles('libs/api/domains/cms/src/lib/generated/contentfulTypes.d.ts', 'apps/**/models/**/*.tsx?', 'libs/**/models/**/*.tsx?', 'apps/**/dto/**/*.tsx?', 'libs/**/dto/**/*.tsx?', 'apps/**/enums/**/*.tsx?', 'libs/**/enums/**/*.tsx?', 'apps/**/queries/**/*.tsx?', 'libs/**/queries/**/*.tsx?', 'apps/air-discount-scheme/web/screens/**/*.tsx', 'apps/air-discount-scheme/web/i18n/withLocale.tsx', 'apps/air-discount-scheme/web/components/AppLayout/AppLayout.tsx', 'apps/air-discount-scheme/web/components/Header/Header.tsx', 'apps/gjafakort/api/src/**/typeDefs.ts')"
```

then it can be referenced with:
```yaml
key: ${{ runner.os }}-${{ *cache_key }}-generated-files
```
(something on these lines)

Comment (darrikonn) [reply .github/workflows/build.yml:491]: Will developers have to edit this variable when they create a new project?

File: .gitignore [hunk 18]
```diff
@@ -50,3 +50,22 @@ helm/**/*.tgz
...
+libs/api/schema/src/lib/schema.d.ts
```
Comment (darrikonn) [inline .gitignore:71]: Can you change this to a wildcard.

File: package.json [hunk 19]
```diff
@@ -34,7 +35,9 @@
     "dep-graph": "nx dep-graph",
     "help": "nx help",
     "storybook:island-ui": "nx run island-ui-storybook:storybook",
-    "ts-node": "node -r ./scripts/ts-node"
+    "ts-node": "ts-node -r tsconfig-paths/register -O '{\"module\":\"commonjs\",\"esModuleInterop\":true}'",
+    "schemas": "test -f libs/api/schema/src/lib/schema.d.ts || echo 'export default () => {}' >> libs/api/schema/src/lib/schema.d.ts && yarn nx run api:contentful-types && yarn affected:schemas",
+    "postinstall": "if [[ ! -z $DOCKER_BUILD ]]; then exit 0; else yarn schemas; fi"
```
Comment (sindrig) [inline package.json:40]: Minor: This will break if DOCKER_BUILD contains spaces or is an empty string. Consider `if [ ! -z ${DOCKER_BUILD+x} ]`

File: scripts/clean-generated-files.sh [hunk 20]
```diff
@@ -0,0 +1,21 @@
...
+rm apps/gjafakort/web/graphql/possibleTypes.json
+rm apps/air-discount-scheme/api.graphql
+rm apps/air-discount-scheme/web/graphql/schema.tsx
+rm apps/air-discount-scheme/web/graphql/possibleTypes.json
+rm apps/air-discount-scheme/backend/src/openapi.yaml
+rm apps/skilavottord/ws/src/app/api.graphql
+rm apps/judicial-system/api/src/openapi.yaml
+rm apps/application-system/api/src/openapi.yaml
+rm apps/reference-backend/src/openapi.yaml
+# rm libs/api/domains/cms/src/lib/generated/contentfulTypes.d.ts
```
Comment (sindrig) [inline scripts/clean-generated-files.sh:20]: ditto - remove commented line?

**Review by sindrig — CHANGES_REQUESTED** [review]
Minor changes but a bit of cleanup required :clap:

File: scripts/clean-generated-files.sh [hunk 21]
```diff
@@ -0,0 +1,21 @@
...
+rm libs/api/schema/src/lib/schema.d.ts
```
Comment (darrikonn) [inline scripts/clean-generated-files.sh:21]: Same with wildcard.
Trying to avoid other developers having to edit this file.

**Review by darrikonn — CHANGES_REQUESTED** [review]

File: package.json (hunk 16 above)
Comment (jeremybarbet) [reply package.json:23]: It's only for local dev

Part of the update on the documentation I wrote:

"However, if you are only focusing on one project (e.g. the api) you can opt-out for others projects to generate their respective files with using an environnement variable into your .env file e.g. `EXCLUDES_SCHEMAS=air-discount-scheme-api,gjafakort-api,skilavottord-ws`"

File: .github/workflows/build.yml (hunk 17 above)
Comment (jeremybarbet) [reply .github/workflows/build.yml:491]: That's a good question and this is the weakest part of the script.

I added the folders (the one that I saw) that trigger the scripts to generate new files (autoSchemeFile/openApi etc). Most of the projects have common folders where it happens e.g. `queries/*`, `models/*`, etc but it's not always the case.

If a new project is created and follows the most common way to create queries/models etc, it will be okay. However, if some files at some specific paths must trigger newly generated files and it's not part of the wildcard it will fail to generate the new schemas.

An alternative would be to never cache theses files on the build steps with the hash key and generate them all the time, but it will increase the build time of ~30sec * number of steps needed

cc/ @eirikurn @sindrig

File: apps/application-system/api/src/app/modules/application/application.module.ts [hunk 22]
```diff
@@ -1,25 +1,34 @@
...
 import { Application } from './application.model'
 import { ApplicationController } from './application.controller'
 import { ApplicationService } from './application.service'
-import { FileStorageModule } from '@island.is/file-storage'
 import { UploadProcessor } from './upload.processor'
 
+let BullModule: DynamicModule
+
+if (process.env.INIT_SCHEMA === 'true') {
+  BullModule = NestBullModule.registerQueueAsync()
```
Comment (eirikurn) [inline apps/application-system/api/src/app/modules/application/application.module.ts:14]: Probably needs to have the same name (upload). I think it's better to put this kind of stuff inside the async useFactory (possibly inline with the @Module declaration like it was):

`
useFactory: () => (inMemory ? {} : { redis: ... }),
`

File: apps/air-discount-scheme/backend/src/app/modules/cache/cache.module.ts [hunk 23]
```diff
@@ -4,9 +4,9 @@ import { createNestJSCache } from '@island.is/cache'
...
 
 const { redis, production } = environment
-
 let CacheModule: DynamicModule
-if (process.env.NODE_ENV === 'test') {
+
+if (process.env.NODE_ENV === 'test' || process.env.INIT_SCHEMA === 'true') {
   CacheModule = NestCacheModule.register()
 } else {
   CacheModule = NestCacheModule.register({
```
Comment (eirikurn) [inline apps/air-discount-scheme/backend/src/app/modules/cache/cache.module.ts:12]: I think `registerAsync` is intended to handle dynamic configuration. cc @darrikonn

File: apps/judicial-system/api/src/buildOpenApi.ts [hunk 24]
```diff
@@ -0,0 +1,11 @@
+import { buildOpenApi } from '@island.is/infra-nest-server'
+
+import { AppModule } from './app/app.module'
+import { openApi } from './openApi'
+
+// TODO - redis
```
Comment (eirikurn) [inline apps/judicial-system/api/src/buildOpenApi.ts:6]: Still valid?

File: package.json (hunk 16 above)
Comment (eirikurn) [reply package.json:23]: This assumes people are using dotenv, but it does not work for normal environment variables, nor for direnv (which some are pushing for).

I think we need to pick one or the other soon.

File: package.json [hunk 25]
```diff
@@ -34,7 +35,9 @@
     "dep-graph": "nx dep-graph",
     "help": "nx help",
     "storybook:island-ui": "nx run island-ui-storybook:storybook",
-    "ts-node": "node -r ./scripts/ts-node"
+    "ts-node": "ts-node -r tsconfig-paths/register -O '{\"module\":\"commonjs\",\"esModuleInterop\":true}'",
```
Comment (eirikurn) [inline package.json:38]: I actually want the `scripts/ts-node` file so I can later require it from js files (like `next.config.js`) and then import shaerd typescript code.

Did you have issues with it?

File: .github/workflows/build.yml (hunk 17 above)
Comment (sindrig) [reply .github/workflows/build.yml:491]: surprise surprise I keep forgetting about yaml variables. we should look into creating a parameter node in the build's root for this.

as for if developers will have to edit this variable. like @jeremybarbet says, if the most common way to create queries/models is used, there will be no problem. and if a developer uses some obscure way, the build will fail and he/she will be faced with the decision to either update this variable (and the .gitignore and removal script probably) or simply use the common way.

File: package.json (hunk 19 above)
Comment (eirikurn) [reply package.json:40]: It's common to check for a CI variable. Does that make sense here?

File: scripts/get-files-touched-by.sh [hunk 26]
```diff
@@ -0,0 +1,11 @@
...
+DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )/.." >/dev/null 2>&1 && pwd )"
+marker=$(mktemp)
+touch $marker -r ${BASH_SOURCE[0]}
+
+CMD=$1
+shift
+
+$CMD "$@" 1>&2
+
+find $DIR/apps $DIR/libs -type f -newer $marker
```
Comment (eirikurn) [inline scripts/get-files-touched-by.sh:11]: Magic! 😮 

Maybe a bash comment is appropriate :)

File: workspace.json [hunk 27]
```diff
@@ -313,6 +313,16 @@
...
           }
         },
+        "init-schema": {
+          "builder": "@nrwl/workspace:run-commands",
+          "options": {
+            "commands": [
+              {
+                "command": "yarn nx run gjafakort-api:codegen"
+              }
+            ]
```
Comment (eirikurn) [inline workspace.json:323]: All of these `run-commands` can be simplified after you merge from master.

File: workspace.json [hunk 28]
```diff
@@ -345,6 +355,23 @@
...
+            "commands": [
+              {
+                "command": "yarn ts-node -P apps/api/tsconfig.app.json apps/api/src/buildSchema.ts"
+              },
+              {
+                "command": "yarn nx run api:codegen"
+              },
+              {
+                "command": "yarn nx run web:codegen"
+              }
```
Comment (eirikurn) [inline workspace.json:371]: Can web have its own `init-schema`? Since we'll end up with a lot of clients to this api.

File: workspace.json [hunk 29]
```diff
@@ -814,6 +810,18 @@
             "cwd": "apps/reference-backend"
           }
         },
+        "init-schema": {
+          "builder": "@nrwl/workspace:run-commands",
+          "options": {
+            "parallel": false,
```
Comment (eirikurn) [inline workspace.json:816]: Parallel not needed.

**Review by eirikurn — DISMISSED** [review]
Awesome job. Minor comments

File: package.json (hunk 19 above)
Comment (sindrig) [reply package.json:40]: The original `CI` variable isn't available inside the docker build. But I wonder if we should name it `CI` for consistency's sake?

File: package.json (hunk 16 above)
Comment (sindrig) [reply package.json:23]: Agreed - can we skip the file check and simply use the environment variable `EXCLUDE_SCHEMAS`? Then the question on dotenv vs direnv can be postponed (or at least not tied to this usecase).

File: apps/application-system/api/src/app/modules/application/application.module.ts (hunk 22 above)
Comment (jeremybarbet) [reply apps/application-system/api/src/app/modules/application/application.module.ts:14]: I don't know why, but with this syntax, the process still keeps running and doesn't stop

File: package.json (hunk 25 above)
Comment (jeremybarbet) [reply package.json:38]: Oh my bad, it wasn't used so I thought I could overwrite it. 

Yes I couldn't pass down the arguments for ts-node e.g. `ts-node -P tsconfig.json file.ts`, probably because it's using `node` instead of `ts-node`.

I can reintroduce another `ts-node-whatever` script specifically for the workspace usage

What do you think?

File: workspace.json (hunk 27 above)
Comment (jeremybarbet) [reply workspace.json:323]: Yes I saw that, hopefully, won't get too many conflicts 😬

File: package.json (hunk 16 above)
Comment (jeremybarbet) [reply package.json:23]: So basically doing:

```suggestion
    "affected:schemas": "nx affected --target=init-schema --parallel --maxParallel=5 --exclude=$EXCLUDES_SCHEMAS",
```

but it won't pick up the exclude from the .env file and/or direnv (I don't know how direnv works)?

File: package.json (hunk 19 above)
Comment (jeremybarbet) [reply package.json]: I renamed it to CI

File: apps/air-discount-scheme/backend/src/app/modules/cache/cache.module.ts (hunk 23 above)
Comment (darrikonn) [reply apps/air-discount-scheme/backend/src/app/modules/cache/cache.module.ts:12]: Ahh ok 👍

File: package.json (hunk 16 above)
Comment (sindrig) [reply package.json:23]: yes. direnv is basically a shellscript where you can export certain environment variables when you enter a directory (or do whatever you want, when you enter a directory)

File: package.json (hunk 25 above)
Comment (eirikurn) [reply package.json:38]: Ah, I see. Makes sense.

Just undelete it for now. No need to create another script in package.json.

File: apps/application-system/api/src/app/modules/application/application.module.ts (hunk 22 above)
Comment (eirikurn) [reply apps/application-system/api/src/app/modules/application/application.module.ts:14]: Ah ... Maybe because of the name parameter, triggering the consumer to start waiting for upload jobs.

File: package.json (hunk 16 above)
Comment (jeremybarbet) [reply package.json:23]: Looks nice. So it's probably better I also remove the section in the document I added about adding this variable inside the .env file until we decided a way to handle these env variables?

File: apps/air-discount-scheme/backend/src/app/modules/cache/cache.module.ts (hunk 23 above)
Comment (jeremybarbet) [reply apps/air-discount-scheme/backend/src/app/modules/cache/cache.module.ts:12]: I shouldn't change it on this PR right?

File: package.json (hunk 16 above)
Comment (sindrig) [reply package.json:23]: Agreed

File: apps/application-system/api/src/app/modules/application/application.module.ts (hunk 22 above)
Comment (jeremybarbet) [reply apps/application-system/api/src/app/modules/application/application.module.ts:14]: Yes adding the name, trigger the upload job.

Should I keep it this way?

File: package.json (hunk 25 above)
Comment (jeremybarbet) [reply package.json:38]: Not sure to understand your comment _Just undelete it for now_. Did you mean _delete_?

File: .github/workflows/build.yml (hunk 17 above)
Comment (sindrig) [reply .github/workflows/build.yml]: Unfortunately, GitHub actions don't support anchors :scream_cat: 
https://github.com/island-is/island.is/issues/1117

File: .gitignore (hunk 18 above)
Comment (jeremybarbet) [reply .gitignore]: I added the changes for both `gitignore` and `clean-generated-files`. However, if someone decides to rename a generated file into `custom-name.d.ts` it will need to be added to the list but it shouldn't be the case

File: scripts/get-files-touched-by.sh [hunk 30]
```diff
@@ -0,0 +1,12 @@
...
+DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )/.." >/dev/null 2>&1 && pwd )"
+marker=$(mktemp)
+touch $marker -r ${BASH_SOURCE[0]}
+
+CMD=$1
+shift
+
+$CMD "$@" 1>&2
+
+# Find all the files paths that are generated out of the yarn schemas script
```
Comment (sindrig) [inline scripts/get-files-touched-by.sh:11]: This file doesn't really have anything to do with `yarn schemas`, although only used for it now. What about `# Find all files that were generated by running the command in argv`?

**Review by sindrig — APPROVED** [review]
LGTM

File: scripts/get-files-touched-by.sh (hunk 30 above)
Comment (jeremybarbet) [reply scripts/get-files-touched-by.sh:11]: You are right I took a shortcut since the script is only used along with yarn schemas but it could be used for other scripts. I will change that, thanks

**Review by sindrig — APPROVED** [review]
:clap:

File: package.json (hunk 25 above)
Comment (eirikurn) [reply package.json:43]: Yeah, I meant to restore the old scripts/ts-node.js file and leave it even though it's not used anymore in `yarn ts-node`.

File: apps/application-system/api/src/app/modules/application/application.module.ts (hunk 22 above)
Comment (eirikurn) [reply apps/application-system/api/src/app/modules/application/application.module.ts:14]: Sure, until we find a better way...

File: apps/air-discount-scheme/backend/src/app/modules/cache/cache.module.ts (hunk 23 above)
Comment (eirikurn) [reply apps/air-discount-scheme/backend/src/app/modules/cache/cache.module.ts:12]: No worries.

File: workspace.json [hunk 31]
```diff
@@ -319,6 +331,18 @@
...
+          "options": {
+            "command": "yarn ts-node -P apps/api/tsconfig.app.json apps/api/src/buildSchema.ts"
+          }
+        },
+        "init-schema": {
+          "builder": "@nrwl/workspace:run-commands",
+          "options": {
+            "command": "yarn nx run api:codegen"
+          }
+        },
```
Comment (eirikurn) [inline workspace.json:345]: What's the difference between `build-schema`, `init-schema` and `codegen` (for developer workflow)?

**Review by eirikurn — APPROVED** [review]
Looks great. Just one clarification question.

File: workspace.json (hunk 31 above)
Comment (jeremybarbet) [reply workspace.json:345]: - `build-schema` is used on the graphql module where autoSchemaFile is defined to generate the `api.graphql` file
- `codegen` is used in projects where a codegen.yml file exists
- `init-schema` is the common script used with the nx affected. For nestjs project it will run in general `build-schema`, for web projects it will run `codegen` scripts and for open api projects it will run `buildOpenApi`.

Comment (jeremybarbet) [reply workspace.json:345]: Actually, `build-schema` is only used once for the API because I run it before all the affected `init-schema`. 

Otherwise, `init-schema` run either or together `buildSchema.ts`/`codegen`/`buildOpenApi.ts` if the project is using Nest modules, web stack or open api.

**Review by eirikurn — APPROVED** [review]

**Review by sindrig — APPROVED** [review]

**Review by darrikonn — APPROVED** [review]

**Review by darrikonn — APPROVED** [review]

**Review by eirikurn — DISMISSED** [review]

**Review by obmagnusson — APPROVED** [review]
LGTM

**Review by sindrig — DISMISSED** [review]

**Review by sindrig — APPROVED** [review]

**Review by sindrig — APPROVED** [review]

**Review by eirikurn — DISMISSED** [review]

**Review by haukurh — DISMISSED** [review]
Aside from one minor comment and the fact that I miss the HelloWorld module already...
This looks good to me

**Review by eirikurn — APPROVED** [review]

**Review by haukurh — DISMISSED** [review]

**Review by gudjong — DISMISSED** [review]

**Review by eirikurn — DISMISSED** [review]

**Review by haukurh — DISMISSED** [review]
